middleware/plugin/plugin.go

```diff
@@ -0,0 +1,311 @@
+package plugin
+
+/*
+ * SPDX-FileCopyrightText: 2023 Siemens AG
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Author: Michael Adler <michael.adler@siemens.com>
+ */
+
+import (
+	"bufio"
+	"errors"
+	"io"
+	"net/http"
+	"os/exec"
+	"sync"
+	"sync/atomic"
+	"syscall"
+	"time"
+
+	"github.com/Southclaws/fault"
+	"github.com/rs/zerolog/log"
+	genPlugin "github.com/siemens/wfx/generated/plugin"
+	"github.com/siemens/wfx/generated/plugin/client"
+	"github.com/siemens/wfx/middleware/logging"
+	"github.com/siemens/wfx/middleware/plugin/ioutil"
+)
+
+type Plugin interface {
+	Name() string
+	Start() error
+	Stop() error
+	IsStopped() bool
+	Wait() error
+	Enqueue(r *http.Request) (*PendingResponse, error)
+}
+
+// FBPlugin is a plugin which communicates using FlatBuffer messages.
+type FBPlugin struct {
+	path string
+
+	queue       chan *QueueEntry
+	queueClosed atomic.Bool
+
+	responses      map[uint64]PendingResponse
+	responsesMutex sync.Mutex
+
+	cookieCounter atomic.Uint64
+	cmd           atomic.Pointer[exec.Cmd]
+	finished      atomic.Bool
+	stopped       atomic.Bool
+
+	errSender   chan error
+	errReceiver chan error
+}
+
+type QueueEntry struct {
+	created  time.Time
+	received time.Time
+	request  *genPlugin.PluginRequestT
+}
+
+func newQueueEntry(request *genPlugin.PluginRequestT) *QueueEntry {
+	return &QueueEntry{
+		created: time.Now(),
+		request: request,
+	}
+}
+
+// PendingResponse represents a response that is pending from the plugin.
+// Use the Await() method to block and wait for the actual response.
+type PendingResponse struct {
+	chResp    chan *genPlugin.PluginResponseT
+	entry     *QueueEntry
+	timeoutFn func()
+}
+
+// NewFBPlugin creates a new plugin instance. In order to start the plugin, call
+// the Start() function.
+func NewFBPlugin(path string) *FBPlugin {
+	return &FBPlugin{path: path}
+}
+
+// Start starts the plugin but does not wait for it to complete.
+//
+// After a successful call to Start the Wait method must be called in
+// order to release associated system resources.
+func (p *FBPlugin) Start() error {
+	if p.finished.Load() || p.stopped.Load() {
+		return errors.New("cannot start a finished plugin")
```

Comment:
The question is whether plugins are allowed to be finished? Could be useful, e.g., for one-time setup things but that can be done in other places as well....

Reply:
Yes, this logic is in place only for the shutdown phase, where we want to stop the plugin and prevent it from being restarted by some other goroutine.

Reply:
I've added a comment in the code